### PR TITLE
Split and limit variables to 999 when using 'IN(...)'

### LINKED ISSUE #2559: Sync fails with 'too many SQL variables' in select
**Describe the bug**
Currently testing with for a user with many patients, sometimes sync fails with error

```
2024-05-31 03:02:29.267 12229-12649 SQLiteLog               org...rtregister.fhircore.mwcoreDev  E  (1) too many SQL variables in "SELECT *
                                                                                                            FROM LocalChangeResourceReferenceEntity
                                                                                                            WHERE localChangeId IN (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?
2024-05-31 03:02:29.269 12229-12374 WM-WorkerWrapper        org...rtregister.fhircore.mwcoreDev  E  Work [ id=1f1e3585-b4ae-48fb-8194-65764141585c, tags={ org.smartregister.fhircore.engine.sync.AppSyncWorker } ] failed because it threw an exception/error
                                                                                                    java.util.concurrent.ExecutionException: android.database.sqlite.SQLiteException: too many SQL variables (code 1 SQLITE_ERROR): , while compiling: SELECT *
                                                                                                            FROM LocalChangeResourceReferenceEntity
2024-05-31 03:02:29.271 12229-12374 WM-WorkerWrapper        org...rtregister.fhircore.mwcoreDev  E          WHERE localChangeId IN (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,
```

**Component**
Engine library

**Additional context**
Checking the LocalEntity table for our app it has around 54k tasks,  it might be related to the above query failing on sync upload

**Would you like to work on the issue?**
Please state if this issue should be assigned to you or who you think could help to solve this issue.

### PR DESCRIPTION
**Pull request by LZRS**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #2559 

**Description**
Chunks the number of variables used in "IN(...)" to 999

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: (Bug fix | Feature | Documentation | Testing | Code health | Builds | Releases | Other)

**Screenshots (if applicable)**

**Checklist**

- [ ] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/CODE_OF_CONDUCT.md).
- [ ] I have read the [Contributing](https://google.github.io/android-fhir/contrib/) page.
- [ ] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [ ] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [ ] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [ ] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [ ] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/LocalChangeDao.kt [hunk 1]
```diff
@@ -498,6 +501,7 @@ internal abstract class LocalChangeDao {
 
   companion object {
     const val DEFAULT_ID_VALUE = 0L
+    const val SQLITE_IN_OPERATOR_VARIABLE_LIMIT = 999
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/db/impl/dao/LocalChangeDao.kt:504]: Add kdoc to explain why this is needed.

File: engine/src/main/java/com/google/android/fhir/db/impl/DatabaseImpl.kt [hunk 2]
```diff
@@ -408,12 +409,14 @@ internal class DatabaseImpl(
   override suspend fun getLocalChangeResourceReferences(
     localChangeIds: List<Long>,
   ): List<LocalChangeResourceReference> {
-    return localChangeDao.getReferencesForLocalChanges(localChangeIds).map {
-      LocalChangeResourceReference(
-        it.localChangeId,
-        it.resourceReferenceValue,
-        it.resourceReferencePath,
-      )
+    return localChangeIds.chunked(SQLITE_IN_OPERATOR_VARIABLE_LIMIT).flatMap { chunk ->
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/db/impl/DatabaseImpl.kt:412]: Add a simple test to check it.

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/LocalChangeDao.kt [hunk 3]
```diff
@@ -412,7 +412,10 @@ internal abstract class LocalChangeDao {
     val updatedReferenceValue = "${updatedResource.resourceType.name}/${updatedResource.logicalId}"
     val referringLocalChangeIds =
       getLocalChangeReferencesWithValue(oldReferenceValue).map { it.localChangeId }.distinct()
-    val referringLocalChanges = getLocalChanges(referringLocalChangeIds)
+    val referringLocalChanges =
+      referringLocalChangeIds.chunked(SQLITE_IN_OPERATOR_VARIABLE_LIMIT).flatMap {
+        getLocalChanges(it)
+      }
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/db/impl/dao/LocalChangeDao.kt:418]: Add a test to check its returning all the local changes (more than limit).

**Review by aditya-07 — DISMISSED** [review]

File: engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt [hunk 4]
```diff
@@ -4086,6 +4088,31 @@ class DatabaseImplTest {
     assertThat(searchedObservations[0].logicalId).isEqualTo(locallyCreatedObservationResourceId)
   }
 
+  @Test
```
Comment (jingtang10) [inline engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt:4091]:
```suggestion
  // https://github.com/google/android-fhir/issues/2559
  @Test
```

File: engine/src/androidTest/java/com/google/android/fhir/db/impl/dao/LocalChangeDaoTest.kt [hunk 5]
```diff
@@ -358,6 +359,38 @@ class LocalChangeDaoTest {
       .isEqualTo(practitionerReference)
   }
 
+  @Test
```
Comment (jingtang10) [inline engine/src/androidTest/java/com/google/android/fhir/db/impl/dao/LocalChangeDaoTest.kt:362]:
```suggestion
  // https://github.com/google/android-fhir/issues/2559
  @Test
```

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/LocalChangeDao.kt [hunk 6]
```diff
@@ -498,6 +501,12 @@ internal abstract class LocalChangeDao {
 
   companion object {
     const val DEFAULT_ID_VALUE = 0L
+
+    /**
+     * Represents SQLite limit on the size of parameters that can be passed in an IN(..) query See
+     * https://issuetracker.google.com/issues/192284727 See https://www.sqlite.org/limits.html
+     */
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/db/impl/dao/LocalChangeDao.kt:508]:
```suggestion
     * See https://github.com/google/android-fhir/issues/2559
     */
```

**Review by jingtang10 — APPROVED** [review]
thanks @LZRS - looks good to me!

**Pull request merged**